### DISCUSSION TITLE
Python version error (Mac)

### DISCUSSION THREAD

**Discussion by jamesblonde007** [Acly/krita-ai-diffusion#740] [General]
After installing Krita and the AI Diffusion plugin, I tried to install a local server. However, it keeps creating the error message that it needs Python v3.8 or later and that I am running 3.7.x. However, as I can confirm in Terminal, I have Python 3.12 installed. Any idea what is going wrong?

**Comment by Acly** [comment]
It searches for the following in order: python3.11, python3.10, python3
The last one is whatever is set as your default python3, which might be 3.7 even if you have 3.12 installed? Although if you're saying that just `python3 --version` in Terminal reports 3.12 it would be strange.

Anyway, last I checked there are some 3rd party packages that don't support 3.12, so it probably won't work.

**Comment by dixhill** [comment]
I'm getting the same error as the OP:
`Exception: Error during installation: typing-extensions requires Python '>=3.8' but the running Python is 3.7.3`
I tried installing Python 3.12, 3.11, and 3.10.  I uninstalled all of those then tried reinstalling 3.11.9 with Homebrew but I keep getting the same error.

I also verified that venv is installed by running python3 -m venv test_env which worked as expected.

What can I try next?

**Comment by Acly** [comment]
1. Make sure you delete the server install folder between attempts (otherwise it tries to continue with an existing install attempt)
2. Make sure `python3.11 --version` finds the correct python in Terminal
3. Check client.log, it should include some info on which Python it finds and from where

**Comment by dixhill** [comment]
Thanks for the quick reply!  Also, I should have mentioned that I'm on a Mac, my specs are below.

1. I deleted this: `~/Library/Application Support/krita/ai_diffusion/server`
2. `python3.11 --version` returns **Python 3.11.9**
3. The client.log says `Using Python: Python 3.7.3` a bunch of times.  What else should I look for?  I would paste the whole thing here but I think if has information I shouldn't reveal online.
4. I also noticed this in the the Server folder: `~Library/Application Support/krita/ai_diffusion/server/venv/lib/python3.7`

System:
MacBook Pro 2019
CPU: 2.3 GHz 8-Core Intel Core i9
GPU: AMD Radeon Pro 5500M 8GB + Intel UHD Graphics 630 1536MB
RAM: 64 GB 2667 MHz DDR4
OS: macOS Ventura v13.6.7

**Comment by Acly** [comment]
> I also noticed this in the the Server folder: ~Library/Application Support/krita/ai_diffusion/server/venv/lib/python3.7

Yea, after a Python version is found, the installer creates a venv, which effectively locks it for installation attempts on the same folder afterwards. That's why deleting it is important if you want it to use a different version.

The installer really just looks for an executable `python3.11` in the environment PATH, if it finds one it will use it. Seems to work for you from the terminal, so I would exepct it to work in Krita too, but I don't know anything about Mac. Maybe terminal uses a different environment.

**Comment by jamesblonde007** [comment]
Ok, I uninstalled Python 3.12.x and installed 3.11.8

running "python3.11 --version" returns "Python 3.11.8". So far so good.

I deleted the plug-in folder from the Krita App Support folder and reinstalled the plug-in (v 1.17.2). When I try to install the local server I get the same "needs Python v3.8 or later " message. 

What am I doing wrong? 

(MacOS 13.6.7, Intel)

**Comment by Acly** [comment]
Is Krita downloaded directly from krita.org, or installed via App Store? Maybe it makes a difference regarding environment.

**Reply by dixhill** [reply]
For me, Krita is downloaded directly from krita.org

**Reply by jamesblonde007** [reply]
> For me, Krita is downloaded directly from krita.org

same

**Comment by dixhill** [comment]
I fixed it! I had to update XCode: [https://xcodereleases.com/](https://xcodereleases.com/)
After doing that, I opened `/usr/bin/python3` (the file that AI-Diffusion is using to find Python) and now it reports `Python 3.9.6` as the default Python version.

Now the server is connected and running!
I'm getting a `Server Execution Error: Invalid Buffer size: 5.00 GB` but at least the old problem is solved.

**Reply by dixhill** [reply]
I fixed the buffer size problem by switching from Automatic to GPU:
Configure Image Diffusion > Performance > Performance Preset \ Device

**Reply by jamesblonde007** [reply]
> I fixed it! I had to update XCode: https://xcodereleases.com/ After doing that, I opened `/usr/bin/python3` (the file that AI-Diffusion is using to find Python) and now it reports `Python 3.9.6` as the default Python version.
> 
> Now the server is connected and running! I'm getting a `Server Execution Error: Invalid Buffer size: 5.00 GB` but at least the old problem is solved.

I tried it but unfortunately it didn't work for me. (I installed XCode 15.2 which is the latest version available for MacOS 16.6.7)

**Reply by dixhill** [reply]
My bad, I should have mentioned I updated to Sonoma 14.5 before I installed XCode.

Before updating to Sonoma, I tried installing **pyenv** and setting the default python to 3.11, but that didn't work.  I used ChatGPT to get instructions and I did a little research to understand what it was telling me to do.

Before that, I temporarily updated the PATH environment variable by setting Python 3.11 as the default in a terminal session.  It resets when you close the terminal session so you have to install AI-Diffusion while the session is still open.  That didn't work either.  I used ChatGPT for this also.

**Reply by Acly** [reply]
If you set PATH in the terminal session the changes work for processes started from that terminal (eg. if  you start Krita from there with something like `open Applications/Krita.app`).

There is probably a way to edit environment variables globally on macOS too.

Sounds like you instead replaced whatever Python it did find before with a newer version, which is also fine. (I suspected 3.9 might be too old for some things, but if it works it works)

**Comment by jamesblonde007** [comment]
Has anyone figured out a solution to this issue yet?

**Comment by jamesblonde007** [comment]
Still trying to get this to work. 

I just updated Krita to the new beta (5.2.3) and updated krita-ai-diffusion  to 1.18.1

Same error. 

In terminal, if I enter 'python3 --version' I get 'Python 3.11.8' which is obviously >3,8

I have tried so many suggestions online and none of them have worked. Any ideas?

**Comment by mariusgarenaux** [comment]
I had the same problem on MacOs Sonoma 14.5, with Krita version 5.2.3 and krita-ai-diffusion 1.21.0. That is, krita-ai-diffusion python venv was installed with a too old version of python, more specifically not the one linked to the command python (neither python3) in PATH variable.

The brute force solution I've found is to reinstall python venv with latest python version : 

- open terminal
- go in the server folder ( for me : /Users/username/Library/Application Support/krita/ai_diffusion/server)
- activate venv : source venv/bin/activate
- keep somewhere a track of all installed python packages : pip freeze > requirements.txt
- delete / rename the folder venv
- create a new venv : python -m venv venv (before, check with 'python --version' that the python version linked to python command is the one you want, for me it is python 3.11.8)
- activate the new venv : source venv/bin/activate
- install packages : pip install -r requirements.txt 

While it is quite an inelegant solution, it seems to have worked for me, **except I now have the 'invalid buffer size' error as @chrisdukes **, but I guess it is related to not using a GPU, and only MPS acceleration.

**Reply by Acly** [reply]
the plugin uses `shutil.which` to find the python executable:
```
import shutil
for cmd in ["python3.11", "python3.10", "python3", "python"]:
  print(cmd, "->", shutil.which(cmd))

```
would be interesting what this prints when executed from terminal python

I rather suspect that Krita is not started with the same environment (PATH) as the terminal for Mac reasons.

**Reply by mariusgarenaux** [reply]
In my python terminal, this outputs (I'm using pyenv, maybe that's part of the issue) : 

```
python3.11 -> /Users/username/.pyenv/versions/3.11.8/bin/python3.11
python3.10 -> None
python3 -> /Users/username/.pyenv/versions/3.11.8/bin/python3
python -> /Users/username/.pyenv/versions/3.11.8/bin/python
```

It seems that Krita uses python from /Users/username/Library instead of the one linked to PATH variable.

**Reply by Acly** [reply]
Presumably because it's not in PATH - environment variables are not the same across all programs. It can be that
* macos doesn't provide the full PATH to Krita 
* the terminal extends the PATH (eg. in a start-up like `.bash_profile` or terminal profile or some other mechanism)
